### DISCUSSION TITLE
How to search for a list of statements [in c++]?

### DISCUSSION THREAD

**Discussion by partner-pierre-vande-laar** [ast-grep/ast-grep#1254] [Q&A]
I would like to find these different instances of ready, set, go in c / c++
```cpp
    ready();
    set();
    go();
    
    ready ( );
    set ( );
    go ( );

    ready(); set(); go();
```

However, I don't get any match:
![image](https://github.com/ast-grep/ast-grep/assets/171016967/3c988619-1bc6-43f9-ba8c-a469c93683d0)

With a single statement, it works as expected:
![image](https://github.com/ast-grep/ast-grep/assets/171016967/88823f0f-b09a-495d-8eb8-86350653e903)

Even when the statements list are not parts of a list (sublist) but really complete lists, they are not found: 
![image](https://github.com/ast-grep/ast-grep/assets/171016967/d06b7ff8-6178-46e4-90ae-dafd519c4ab2)

Only when all are compound statements the expected matches are found.
![image](https://github.com/ast-grep/ast-grep/assets/171016967/abd958e9-dce2-45ce-a2c0-81f8a95b0b8a)

Hence, my question
**Can't ast-grep handle lists of statements?**

**Comment by partner-pierre-vande-laar** [comment]
Made a copy of this question in Discord (as was recommended by a general message).

**Comment by HerringtonDarkholme** [comment]
No, ast-grep's pattern does not handle a list of ast-nodes.

You can either use
1.  find the parent node and rewriter https://ast-grep.github.io/reference/yaml/rewriter.html
2. use ast-grep js/py API https://ast-grep.github.io/guide/api-usage.html

**Reply by pjljvandelaar** [reply]
Thanks for the info.
It is not what I would have liked, but it is clear.

**Reply by pjljvandelaar** [reply]
Note that a list of ast-nodes is very useful.
See e.g., https://github.com/TNO/Rewriters-Ada/blob/736dbf041188f3ed22423f91467df0d6358b5197/src/predefined_rewriters_prefer_quantified_expressions.ads#L73
for a few examples using Ada.

**Reply by HerringtonDarkholme** [reply]
You can use other tools meeting your need.

**Reply by pjljvandelaar** [reply]
OK, I will look for better alternatives!
